gemrb/core/Scriptable/Actor.cpp

```diff
@@ -5436,22 +5433,57 @@ int Actor::GetEncumbranceFactor(bool feedback) const
 	return 123456789; // large enough to round to 0 when used as a divisor
 }
 
-// NOTE: for ini-based speed users this will only update their encumbrance, speed will be 0
 int Actor::CalculateSpeed(bool feedback)
 {
-	int speed = GetStat(IE_MOVEMENTRATE);
+	if (core->HasFeature(GF_RESDATA_INI)) {
+		CalculateSpeedFromINI(feedback);
+	} else {
+		CalculateSpeedFromRate(feedback);
+	};
+	return speed;
+}
 
+// NOTE: for ini-based speed users this will only update their encumbrance, speed will be 0
+void Actor::CalculateSpeedFromRate(bool feedback)
+{
+	int movementRate = GetStat(IE_MOVEMENTRATE);
+	int encumbranceFactor = GetEncumbranceFactor(feedback);
 	if (BaseStats[IE_EA] > EA_GOODCUTOFF && !third) {
 		// cheating bastards (drow in ar2401 for example)
-		return speed;
+	} else {
+		movementRate /= encumbranceFactor;
 	}
+	if (movementRate) {
+		speed = 1500 / movementRate;
+	} else {
+		speed = 0;
+	}
+}
 
-	inventory.CalculateWeight();
-	int encumbrance = inventory.GetWeight();
+void Actor::CalculateSpeedFromINI(bool feedback)
+{
 	int encumbranceFactor = GetEncumbranceFactor(feedback);
-	SetStat(IE_ENCUMBRANCE, encumbrance, false);
-
-	return speed / encumbranceFactor;
+	ieDword animid = BaseStats[IE_ANIMATION_ID];
+	if (core->HasFeature(GF_ONE_BYTE_ANIMID)) {
+		animid = animid & 0xff;
+	}
+	if (animid < (ieDword)CharAnimations::GetAvatarsCount()) {
```

Comment:
Since we're cleaning up, let's just make this an assert, since it should always be true. Will reduce the nesting by one level as well.